src/apps/polymer/IncentivizedPolymerEscrow.sol

```diff
@@ -0,0 +1,175 @@
+            bytes32(abi.encodePacked(packet.srcPortAddr)),
+            sourceImplementationIdentifier,
+            packet.appData,
+            feeRecipitent,
+            gasLimit
+        );
+
+        // Send ack:
+        return AckPacket({success: true, data: receiveAck});
+    }
+
+    // The escrow manages acks, so any message can be directly provided to _handleAck.
+    function onUniversalAcknowledgement(bytes32 channelId, UniversalPacket calldata packet, AckPacket calldata ack)
+        external
+        onlyIbcMw
+    {
+        uint256 gasLimit = gasleft();
+        bytes32 feeRecipitent = bytes32(uint256(uint160(tx.origin)));
+
+        bytes calldata rawMessage = ack.data;
+        bytes32 chainIdentifier = _getChannelId(channelId);
+        bytes memory destinationImplementationIdentifier = abi.encodePacked(packet.destPortAddr);
+
+        isVerifiedMessageHash[keccak256(rawMessage)] = VerifiedMessageHashContext({
+            chainIdentifier: chainIdentifier,
+            implementationIdentifier: destinationImplementationIdentifier
+        });
+        _handleAck(chainIdentifier, destinationImplementationIdentifier, rawMessage, feeRecipitent, gasLimit);
+    }
+
+    // For timeouts, we need to construct the message.
+    function onTimeoutUniversalPacket(bytes32 channelId, UniversalPacket calldata packet) external onlyIbcMw {
+        uint256 gasLimit = gasleft();
+        bytes32 feeRecipitent = bytes32(uint256(uint160(tx.origin)));
+
+        bytes calldata rawMessage = packet.appData;
+        _handleTimeout(_getDestChainId(channelId), rawMessage, feeRecipitent, gasLimit);
+    }
+
+    // * Send to messaging_protocol
+    function _sendPacket(
+        bytes32 destinationChainIdentifier,
+        bytes memory, /* destinationImplementation */
+        bytes memory message
+    ) internal override returns (uint128 costOfsendPacketInNativeToken) {
+        // call `setChannelId` to set the channelId for the destination chain before calling this function.
+        bytes32 channelId = _getChannelId(destinationChainIdentifier);
+        // set timeoutTimestamp to 1 day from now. It's the dest chain's block time in nanoseconds since the epoch.
+        uint64 timeoutTimestamp = uint64(block.timestamp + 1 days) * 1e9;
+        IbcUniversalPacketSender(mw).sendUniversalPacket(
+            channelId, destinationChainIdentifier, message, timeoutTimestamp
+        );
+        return 0;
+    }
+
+    // ChannelIds are always from the running chain's perspective.
+    // Each universal channel/channelId represents a directional path from the running chain to a destination chain.
+    // Universal ChannelIds should _destChainIdToChannelIdd from the Polymer registry.
+    // Although everyone is free to establish their own channels, they're not "officially" vetted until they're in the Polymer registry.
+    mapping(bytes32 => bytes32) _destChainChannelIds;
+    mapping(bytes32 => bytes32) _channelIdToDestChainId;
```

Comment:
Is it better to fetch these directly from the Polymer registry rather than storing them here?

Reply:
Following our discussion, treat channelId as chainIdentifier. So get rid of stored registry in this Escrow contract